src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -309,5 +309,60 @@ public void PresentDifferentChains_AlternativeChainWithMoreChainWorkShouldAlways
+        /// first checkpoint with a prolongation that does not match the 2nd checkpoint. Exception should be thrown.
+        /// </summary>
+        [Fact]
+        public void ChainHasTwoCheckPoints_ChainConveringOnlyFirstCheckPointIsPresented_ChainIsDiscardedUpUntilFirstCheckpoint()
+        {
+            // Chain header tree setup.
+            const int initialChainSize = 2;
+            const int currentChainExtension = 6;
+            var ctx = new TestContext();
+            ChainedHeaderTree cht = ctx.CreateChainedHeaderTree();
+            ChainedHeader initialChainTip = ctx.ExtendAChain(initialChainSize); // ie. h1=h2
+            cht.Initialize(initialChainTip, true); 
+            ChainedHeader extendedChainTip = ctx.ExtendAChain(currentChainExtension, initialChainTip); // ie. h1=h2=h3=h4=h5=h6=h7=h8
+            ctx.ConsensusSettings.UseCheckpoints = true;
+            List<BlockHeader> listOfCurrentChainHeaders = ctx.ChainedHeaderToList(extendedChainTip, initialChainSize + currentChainExtension);
+
+            // Setup two known checkpoints at header 4 and 7.
+            // Example: h1=h2=h3=(h4)=h5=h6=(h7)=h8.
+            const int firstCheckpointHeight = 4;
+            const int secondCheckpointHeight = 7;
+            var checkpoint1 = new CheckpointInfo(listOfCurrentChainHeaders[firstCheckpointHeight - 1].GetHash());
+            var checkpoint2 = new CheckpointInfo(listOfCurrentChainHeaders[secondCheckpointHeight - 1].GetHash());
+            ctx.CheckpointsMock
+               .Setup(c => c.GetCheckpoint(firstCheckpointHeight))
+               .Returns(checkpoint1);
+            ctx.CheckpointsMock
+               .Setup(c => c.GetCheckpoint(secondCheckpointHeight))
+               .Returns(checkpoint2);
+            ctx.CheckpointsMock
+               .Setup(c => c.GetCheckpoint(It.IsNotIn(firstCheckpointHeight, secondCheckpointHeight)))
+               .Returns((CheckpointInfo)null);
+            ctx.CheckpointsMock
+                .Setup(c => c.GetLastCheckpointHeight())
+                .Returns(secondCheckpointHeight);
+
+            // Setup new chain that only covers first checkpoint but doesn't cover second checkpoint.
+            // Example: h1=h2=h3=(h4)=h5=h6=x7=x8=x9=x10.
+            const int newChainExtension = 4;
+            extendedChainTip = extendedChainTip.Previous; // walk back to block 6
+            extendedChainTip = extendedChainTip.Previous;
+            extendedChainTip = ctx.ExtendAChain(newChainExtension, extendedChainTip); 
+            List<BlockHeader> listOfNewChainHeaders = ctx.ChainedHeaderToList(extendedChainTip, extendedChainTip.Height);
+            
+            // New chain is presented by peer 1.
+            // InvalidHeaderException should be thrown.
+            Action connectAction = () =>
+            {
+                cht.ConnectNewHeaders(1, listOfNewChainHeaders);
```

Comment:
We wanted the peer to 1st present such a chain that covers first checkpoint only. So in this case it would be for example up to block number five. There should be no problem with such presentation. And then the peer will try to prolong using blocks 6 to 10 which will violate the checkpoint.

So the test is correct except that we just want to split the presentation of headers into two separated steps so that we can distinguish between the first attempt which is correct and should be successful from the other attempt.

This should be only minor modification to the code. Otherwise it looks good.

Reply:
Once I present first part, do I need to set consensus tip after that before presenting second part or not?

Reply:
No, you don't need to do anything with consensus tip.